test/gtest/ucp/test_ucp_worker.cc

```diff
@@ -0,0 +1,223 @@
+
+        m_flush_comps.clear();
+        m_pending_reqs.clear();
+    }
+
+    void test_worker_discard(void *ep_flush_func,
+                             void *ep_pending_add_func,
+                             unsigned ep_count = 8,
+                             unsigned wireup_ep_count = 0,
+                             unsigned wireup_aux_ep_count = 0) {
+        uct_iface_ops_t ops                  = {0};
+        unsigned created_wireup_aux_ep_count = 0;
+        unsigned total_ep_count              = ep_count + wireup_aux_ep_count;
+        uct_iface_t iface;
+        std::vector<uct_ep_t> eps(total_ep_count);
+        std::vector<uct_ep_h> wireup_eps(wireup_ep_count);
+        ucp_ep_t ucp_ep;
+
+        ASSERT_LE(wireup_ep_count, ep_count);
+        ASSERT_LE(wireup_aux_ep_count, wireup_ep_count);
+
+        ucp_ep.worker = sender().worker();
+
+        ops.ep_flush       = (uct_ep_flush_func_t)ep_flush_func;
+        ops.ep_pending_add = (uct_ep_pending_add_func_t)ep_pending_add_func;
+        ops.ep_destroy     = ep_destroy_func;
+        iface.ops          = ops;
+
+        for (unsigned i = 0; i < ep_count; i++) {
+            uct_ep_h discard_ep;
+
+            eps[i].iface = &iface;
+            m_created_ep_count++;
+
+            if (i < wireup_ep_count) {
+                ucs_status_t status = ucp_wireup_ep_create(&ucp_ep, &discard_ep);
+                EXPECT_UCS_OK(status);
+
+                wireup_eps.push_back(discard_ep);
+                ucp_wireup_ep_set_next_ep(discard_ep, &eps[i]);
+
+                if (i < wireup_aux_ep_count) {
+                    eps[ep_count + created_wireup_aux_ep_count].iface = &iface;
+
+                    ucp_wireup_ep_t *wireup_ep = ucp_wireup_ep(discard_ep);
+                    wireup_ep->aux_ep          = &eps[ep_count +
+                                                      created_wireup_aux_ep_count];
+
+                    created_wireup_aux_ep_count++;
+                    m_created_ep_count++;
+                }
+            } else {
+                discard_ep = &eps[i];
+            }
+
+            EXPECT_LE(m_created_ep_count, total_ep_count);
+
+            ucp_worker_discard_uct_ep(sender().worker(), discard_ep,
+                                      UCT_FLUSH_FLAG_LOCAL);
+        }
+
+        do {
+            progress();
+
+            if (!m_flush_comps.empty()) {
+                uct_completion_t *comp = m_flush_comps.back();
+
+                m_flush_comps.pop_back();
+                uct_invoke_completion(comp, UCS_OK);
+            }
+
+            if (!m_pending_reqs.empty()) {
+                uct_pending_req_t *req = m_pending_reqs.back();
+
+                ucs_status_t status = req->func(req);
+                if (status == UCS_OK) {
+                    m_pending_reqs.pop_back();
+                } else {
+                    EXPECT_EQ(UCS_INPROGRESS, status);
+                }
+            }
+        } while (m_destroyed_ep_count < m_created_ep_count);
```

Comment:
can't we just call ucp_worker_flush() and afterwards expect all eps were discarded?

Reply:
but we still need to call flush_completion or peding_req's cb
so, this loop is needed anyway

Reply:
so maybe do nonblocking flush and call these in the wait loop?
we want to test that when worker is flushed, all eps are destroyed..

Reply:
> so maybe do nonblocking flush and call these in the wait loop?

yes, good point
done

> we want to test that when worker is flushed, all eps are destroyed..

yes, I didn't think about this